discover-cpu-partitioning.py

```diff
@@ -0,0 +1,173 @@
+            kernel_cli = fh.readline().rstrip()
+            if t_global.args.debug:
+                debug("get_rcu_nocbs: kernel_cli: %s" % (kernel_cli))
+
+            # find 'rcu_nocbs=<cpu-list>' if it exists in the kernel command line
+            match = re.search(r"rcu_nocbs=[0-9\-,]+", kernel_cli)
+            if match:
+                partition = match.group(0).partition('=')
+                if partition[1] == partition[2] == '':
+                    # this implies that 'rcu_nocbs' exists in the
+                    # kernel command line but without a <cpu-list>
+                    pass
+                else:
+                    if t_global.args.debug:
+                        debug("get_rcu_nocbs: found rcu_nocbs=%s" % (partition[2]))
+                    rcu_nocbs = system_cpu_topology.parse_cpu_list(partition[2])
+    else:
+        raise FileNotFoundError("get_rcu_nocbs: Could not find '%s'" % (input_file))
+
+    return(rcu_nocbs)
+
+def get_pid_cpus_allowed(pid):
+    path = Path('/proc/' + pid + '/status')
+    cpus_allowed = []
+
+    if path.exists() and path.is_file():
+        with path.open() as fh:
+            contents = fh.readlines()
+            for line in contents:
+                match = re.search(r"Cpus_allowed_list", line)
+                if match:
+                    line = line.rstrip()
+                    if t_global.args.debug:
+                        debug("get_pid_cpus_allowed: found '%s'" % (line))
+                    pieces = line.split()
+                    if len(pieces) != 2:
+                        raise AttributeError("get_pid_cpus_allowed: Could not extract cpus_allowed from '%s'" % (line))
+                    else:
+                        cpus_allowed = system_cpu_topology.parse_cpu_list(pieces[1])
+                        if t_global.args.debug:
+                            debug("get_pid_cpus_allowed: extracted cpus_allowed=%s" % (cpus_allowed))
+    else:
+        raise FileNotFoundError("get_pid_cpus_allowed: Requested pid '%s' probably does not exist" % (pid))
+
+    if len(cpus_allowed) == 0:
+        raise AttributeError("get_pid_cpus_allowed: failed to determine cpus_allowed")
+
+    return(cpus_allowed)
+
+def main():
+    process_options()
+
+    if t_global.args.environment == "process":
+        system_cpus = system_cpu_topology(debug = t_global.args.debug)
+
+        all_cpus = system_cpus.get_all_cpus()
+        print("all cpus: %s" % (all_cpus))
+        print("all cpus: %s" % (system_cpu_topology.formatted_cpu_list(all_cpus)))
+        print("all cpus: %d" % (len(all_cpus)))
+        print("")
+        online_cpus = system_cpus.get_online_cpus()
+        print("online cpus: %s" % (online_cpus))
+        print("online cpus: %s" % (system_cpu_topology.formatted_cpu_list(online_cpus)))
+        print("online cpus: %d" % (len(online_cpus)))
+        print("")
+        rcu_nocbs_cpus = get_rcu_nocbs()
+        print("rcu_nocbs: %s" % (rcu_nocbs_cpus))
```

Comment:
I wonder if we should also consider (1) systemd's CPUAffinity or (2) taskset -pc 1 or taskset -pc $$, if this is not a container.  It's possible we could have lesser isolation, or user-space only isolation.

Reply:
I think so.  This is extendable and just a first pass so we can certainly add on additional methods.